lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -774,8 +817,62 @@ void FIRRTLTypesLowering::getAllBundleLowerings(
       .Default([&](auto) {});
 }
 
-Identifier FIRRTLTypesLowering::getArgAttrName(unsigned newArgNumber) {
+Identifier TypeLoweringVisitor::getArgAttrName(unsigned newArgNumber) {
   SmallString<16> argNameBuffer;
   mlir::impl::getArgAttrName(newArgNumber, argNameBuffer);
-  return Identifier::get(argNameBuffer, &getContext());
+  return Identifier::get(argNameBuffer, context);
 }
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerTypesPass : public LowerFIRRTLTypesBase<LowerTypesPass> {
+  void runOnOperation() override;
+
+private:
+  void runAsync();
+  void runSync();
+};
+} // end anonymous namespace
+
+void LowerTypesPass::runAsync() {
+  // Collect the operations to iterate in a vector. We can't use parallelFor
+  // with the regular op list, since it requires a RandomAccessIterator. This
+  // also lets us use parallelForEachN, which means we don't have to
+  // llvm::enumerate the ops with their index. TODO(mlir): There should really
+  // be a way to do this without collecting the operations first.
+  auto &body = getOperation().getBody()->getOperations();
+  std::vector<Operation *> ops;
+  llvm::for_each(body, [&](Operation &op) { ops.push_back(&op); });
+
+  mlir::ParallelDiagnosticHandler diagHandler(&getContext());
+  llvm::parallelForEachN(0, ops.size(), [&](auto index) {
+    // Notify the handler the op index and then perform lowering.
+    diagHandler.setOrderIDForThread(index);
+    TypeLoweringVisitor(&getContext()).lowerModule(ops[index]);
+    diagHandler.eraseOrderIDForThread();
+  });
+}
```

Comment:
I wasn't able to find a way to do this without collecting the ops first.  What I wanted to do here was:
```c++
void LowerTypesPass::runAsync() {
  auto &body = getOperation().getBody()->getOperations();

  mlir::ParallelDiagnosticHandler diagHandler(&getContext());
  llvm::parallelForEach(llvm::enumerate(body), [&](auto indexAndOp) {
    // Notify the handler the op index and then perform lowering.
    diagHandler.setOrderIDForThread(std::get<0>(indexAndOp));
    TypeLoweringVisitor(&getContext()).lowerModule(std::get<1>(indexAndOp));
    diagHandler.eraseOrderIDForThread();
  });
}
```
This didn't work, due to enumerate not returning a random access iterator.  There might be a way to fix this upstream, but I'm not sure of all the details yet.  Any better ideas how to implement this?